tests/js/requestAware/test-request-aware-router.js

```diff
@@ -0,0 +1,111 @@
+// Sample router module to test the `gpii.express.requestAware` abstract grade.
+//
+// It is only meant for testing purposes and should not be extended.  It accepts a single parameter, `req.query.action`,
+// and executes various scenarios based on that action.
+"use strict";
+var fluid = fluid || require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+
+fluid.registerNamespace("gpii.express.tests.requestAware.request");
+gpii.express.tests.requestAware.request.recordTime = function (that) {
+    that.time = new Date();
+};
+
+gpii.express.tests.requestAware.request.sendInstrumentedResponse = function (that, statusCode, body) {
+    var instrumentedBody = fluid.copy(body);
+    instrumentedBody.time = that.time;
+    instrumentedBody.action = that.request.query.action;
+
+    // Send the instrumented response using the standard function
+    gpii.express.requestAware.sendResponse(that, statusCode, instrumentedBody);
+};
+
+fluid.defaults("gpii.express.tests.requestAware.request", {
+    gradeNames: ["gpii.express.requestAware", "autoInit"],
+    members: {
+        time: null
+    },
+    listeners: {
+        "onCreate.recordTime": {
+            funcName: "gpii.express.tests.requestAware.request.recordTime",
+            args:     ["{that}"]
+        }
+    },
+    invokers: {
+        sendResponse: {
+            funcName: "gpii.express.tests.requestAware.request.sendInstrumentedResponse",
+            args: ["{that}", "{arguments}.0", "{arguments}.1"]
+        }
+    }
+});
+
+fluid.registerNamespace("gpii.express.tests.requestAware.router");
+gpii.express.tests.requestAware.router.getRouter = function (that) {
+    return function (req, res) {
+        if (!req.query) {
+            fluid.fail("Can't continue without the req.query object created by query parsing middleware.");
+        }
+
+        // Simulate a timeout
+        if (req.query.action === "timeout") {
+            that.events.onNewTimeoutRequest.fire(req, res);
+        }
+        // Simulate an upstream request and respond when the time is right.
+        else {
+            that.events.onNewDelayedRequest.fire(req, res);
+        }
+    };
+};
+
+fluid.defaults("gpii.express.tests.requestAware.router", {
+    gradeNames: ["gpii.express.router", "autoInit"],
+    method:     "get",
+    path:       "/requestAware",
+    events: {
+        "onNewDelayedRequest": null,
+        "onNewTimeoutRequest": null
+    },
+    components: {
+        urlencoded: {
+            type: "gpii.express.middleware.bodyparser.urlencoded"
+        }
+    },
+    dynamicComponents: {
```

Comment:
From our chat discussion:

Do you have a link for using dynamic options in the way you describe, with a gradeName coming from the component options? Tony @ 5:19	

Unfortunately not antranigbasman@googlemail.com @ 5:20
I imagine the gradeName would come from the event arguments 5:20
Well, perhaps not 5:21

can I use an IoC options reference in the type field? Tony @ 5:21	

I guess if we really went for "one requestAware per router" it would indeed come from the options antranigbasman@googlemail.com @ 5:21
Yes, you can 5:21


great, I will give that a go. Tony @ 5:22	

But the type field itself would probably be taken up with "requestAware" itself antranigbasman@googlemail.com @ 5:22
And the user can get to supply a list of additional grades that go into gradeNames 5:22

I have no idea how to construct any of that. Tony @ 5:22
an expander around the gradeNames? 5:22

No, just the same antranigbasman@googlemail.com @ 5:22
An IoC reference to a field in the options